compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp

```diff
@@ -444,6 +449,70 @@ Attribute EncodingAttr::cloneWithLayouts(ArrayRef<Attribute> layouts) const {
   return LayoutAttr::get(ctx, ArrayAttr::get(ctx, layouts));
 }
 
+/// Helper to compute new shape when bitcasting a tensor to a different element
+/// type. Returns failure if the tensor is invalid or bits don't divide evenly.
+/// On success, returns the new shape with the last dimension adjusted.
+static FailureOr<SmallVector<int64_t>>
+computeBitcastShape(RankedTensorType tensorType, Type newElementType) {
+  ArrayRef<int64_t> shape = tensorType.getShape();
+  if (shape.empty()) {
+    return failure();
+  }
+
+  unsigned oldBitWidth =
+      IREE::Util::getTypeBitWidth(tensorType.getElementType());
+  unsigned newBitWidth = IREE::Util::getTypeBitWidth(newElementType);
+
+  // If bit widths are the same, shape doesn't change.
+  if (oldBitWidth == newBitWidth) {
+    return SmallVector<int64_t>(shape);
+  }
+
+  int64_t lastDim = shape.back();
```

Comment:
Two questions:

1. Do we return failure if the last dim is dynamic?
2. Do we return the same shape if the last dim is dynamic && oldBitWidth is a multiple of newBitWidth?

We can leave some TODO items if we don't want to expand the support broadly, since this is new to us.